File: zig/src/finished/lut.zig
const std = @import("std");
const utils = @import("utils.zig");

const print = std.debug.print;

// TODO: this is the experimental change
// 0.1 increment -> comptime is faster, fits in the L1 cache
// 0.01 increment -> runtime is faster, does not fit in the L1 cache
const increment: comptime_float = 2;
const TEST_SIZE: usize = 500;
const degrees: comptime_float = 360;
const steps: comptime_int = @intFromFloat(degrees / increment);

fn generateLUT() [steps]f64 {
    @setEvalBranchQuota(1000000);
    // table empty, but enough to hold all the steps
    var table: [steps]f64 = undefined;

    for (&table, 0..) |*item, i| {
        const result: f64 = @as(f64, @floatFromInt(i)) * increment;
        item.* = @sin(result) + @cos(result);
    }

    return table;
}

pub fn roundToIncrement(value: f64, inc: f64) f64 {
    const multiplier = 1.0 / inc;
    return std.math.round(value * multiplier) / multiplier;
}

fn generateTestCases(io: anytype, path: []const u8) !void {
    var file = try std.Io.Dir.cwd().createFile(io, path, .{ .truncate = true });
    defer file.close(io);

    var file_writer = file.writer(io, &.{});

    var prng = std.Random.DefaultPrng.init(12345);

    for (0..TEST_SIZE) |_| {
        // steps is the max index that you can access in the LUT
        // just store a random index to look up
        const val = @abs(@mod(prng.random().int(i64), steps));

        try file_writer.interface.print("{d}\n", .{val});

        // std.debug.print("{d},", .{val});
    }
}

pub fn main(init: std.process.Init) !void {
    _ = try generateTestCases(init.io, "lookup.txt");
    const test_cases = try utils.readArrayFromFile(TEST_SIZE, i64, init.io, "lookup.txt");

    const myLut = comptime generateLUT();

    var sum_comp: f64 = 0;

    const io = init.io;

    var start_time = std.Io.Clock.now(.awake, io);

    for (test_cases) |num| {
        //for (0..TEST_SIZE) |num| {
        // std.debug.print("input: {}, output: {}\n", .{ num, myLut[@intCast(num)] });
        sum_comp += myLut[@intCast(num)];
    }

    var end_time = std.Io.Clock.now(.awake, io);
    const duration_comp = start_time.durationTo(end_time);

    var sum_run: f64 = 0;

    start_time = std.Io.Clock.now(.awake, io);

    // dynamic
    for (test_cases) |num| {
        // for (0..TEST_SIZE) |num| {
        const float_num = @as(f64, @floatFromInt(num));
        const angle = float_num * @as(f64, increment);

        sum_run += @sin(angle) + @cos(angle);
    }

    end_time = std.Io.Clock.now(.awake, io);
    const duration_run = start_time.durationTo(end_time);
    std.debug.print("Runtime processed in: {} ns\n", .{duration_run.toNanoseconds()});
    std.debug.print("Comptime processed in: {} ns\n", .{duration_comp.toNanoseconds()});
    print("Sum comp: {d}\n", .{sum_comp});
    print("Sum run: {d}\n", .{sum_run});
}

File: cpp/src/lut.cpp
#include "utils.cpp"
#include <array>
#include <chrono>
#include <cmath>
#include <cstddef>
#include <print>

constexpr double INCREMENT = 0.01;
constexpr size_t TEST_SIZE = 500;
constexpr double DEGREES = 360;
constexpr int STEPS = static_cast<size_t>(DEGREES / INCREMENT);

constexpr std::array<double, STEPS> generate_lut() {
  std::array<double, STEPS> table;

  // double ref ? eh
  for (int i = 0; i < STEPS; i++) {
    const double result = i * INCREMENT;

    table[i] = __builtin_sin(result) + __builtin_cos(result);
  }

  return table;
}

int main() {

  const auto test_cases = utils::read_array_from_file<TEST_SIZE>("lookup.txt");

  // std::print("This is the array read from the file: ");
  //  utils::printArray<TEST_SIZE>(test_cases);

  // Must use C++ 26 (clang is still registering as an error, but it compiles)
  constexpr auto MY_LUT = generate_lut();

  std::print(
      "LUT size: {}, increment: {}, testSize: {}, degrees: {}, steps: {}\n",
      MY_LUT.size(), INCREMENT, TEST_SIZE, DEGREES, STEPS);

  // A:
  // double sum_comp = 0;

  auto start_time = std::chrono::steady_clock::now();

  /*

  // A:

  for (const auto &num : test_cases) {
    const size_t idx = (num / INCREMENT);

    sum_comp += MY_LUT[idx];

    //  std::print("curr comptime test case: {}, lut value: {}\n", num,
    //  myLut[num]);
  }

  */
  double sum_comp = 0.0;

  for (const auto &num : test_cases) {
    // num is strictly a raw test case value here
    const size_t idx = static_cast<size_t>(num / INCREMENT);

    sum_comp += MY_LUT[idx];
  }

  auto end_time = std::chrono::steady_clock::now();
  const auto duration_comp = end_time - start_time;

  double sum_run = 0;

  start_time = std::chrono::steady_clock::now();

  for (const auto &num : test_cases) {

    double float_num = static_cast<double>(num);
    sum_run += std::sin(float_num) + std::cos(float_num);
  }

  end_time = std::chrono::steady_clock::now();
  const auto duration_run = (end_time - start_time).count();

  std::print("Runtime processed in: {} ns\n", duration_run);
  std::print("Comptime processed in: {} ns\n", duration_comp);
  std::print("Sum comp: {}\n", sum_comp);
  std::print("Sum run: {}\n", sum_run);

  return 0;
}
